xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -1071,21 +852,30 @@ public void run() {
    */
   private static class RoutingConfig {
     private final long fallbackTimeoutNano;
-    private final List<Route> routes;
-    private final boolean applyFaultInjection;
-    @Nullable
-    private final HttpFault faultConfig;
+    final List<Route> routes;
+    // Null if HttpFilter is not supported.
+    @Nullable final List<NamedFilter> filterChain;
+    final Map<String, FilterConfig> virtualHostOverrideConfig;
 
-    private static final RoutingConfig empty =
-        new RoutingConfig(0L, Collections.<Route>emptyList(), false, null);
+    private static RoutingConfig empty = new RoutingConfig(
+        0L, Collections.<Route>emptyList(), null, Collections.<String, FilterConfig>emptyMap());
 
     private RoutingConfig(
-        long fallbackTimeoutNano, List<Route> routes, boolean applyFaultInjection,
-        HttpFault faultConfig) {
+        long fallbackTimeoutNano, List<Route> routes, @Nullable List<NamedFilter> filterChain,
+        Map<String, FilterConfig> virtualHostOverrideConfig) {
       this.fallbackTimeoutNano = fallbackTimeoutNano;
       this.routes = routes;
-      this.applyFaultInjection = applyFaultInjection;
-      this.faultConfig = faultConfig;
+      if (filterChain == null) {
+        this.filterChain = null;
+      } else {
+        checkArgument(!filterChain.isEmpty(), "filterChain is empty");
+        this.filterChain = Collections.unmodifiableList(filterChain);
+      }
+      this.virtualHostOverrideConfig = Collections.unmodifiableMap(virtualHostOverrideConfig);
+    }
+
+    private boolean hasLameFilter() {
```

Comment:
I don't see any usefulness of this method, especially compared with what you previously had (aka, a local variable when populating the filter chain). What I meant was, you do not need to populate the filter chain to see if a router filter exists (aka, code in `updateRoutes()`), just peek the last one. If it's not a router filter, insert a LAME.

Reply:
Peeking the last one is not sufficient as the router may be present but is not the last one:
(https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md#the-router-filter)

> In addition, if the router filter is present, any filters in the list after the router filter will not actually be used, although their configurations will still be validated when the config is received from the xDS server.

Reply:
> any filters in the list after the router filter will not actually be used

If it's a router filter presents in the middle, anything after it doesn't matter.

Reply:
Not sure I understood what you mean correctly, removed the method.